poseidon2-starky/src/generation.rs

```diff
@@ -59,6 +136,28 @@ pub fn generate_poseidon2_trace<F: RichField>(step_rows: Vec<Row<F>>) -> [Vec<F>
         for j in 0..STATE_SIZE {
             trace[COL_OUTPUT_START + j][i] = outputs[j];
         }
+
+        // Generate the full round states
+        let first_full_round_state = generate_1st_full_round_state(&row.preimage);
+        let partial_round_state = generate_partial_round_state(
+            first_full_round_state.last().unwrap().try_into().unwrap(),
+        );
+        let second_full_round_state =
+            generate_2st_full_round_state(partial_round_state.last().unwrap().try_into().unwrap());
```

Comment:
just a thought, what if we keep state in scalar and convert them to field only after second_full_round_state is generated in terms of scalar? could we same field_to_scalar and scalar_to_field in generate_..() functions?

Reply:
We could save two field_to_scalar_vec conversions, but it might slightly compromise readability, in my opinion.

You can't convert them to fields only after the second_full_round. This is because we need field outputs from all three rounds to decrease the degree of constraint.